ticdc/ticdc-craft.md

```diff
@@ -0,0 +1,385 @@
+* [Resolved Event](#resolved-event): Represents a special time point before which the Event received is complete.
+
+## Restrictions
+
+* In most cases, the Row Changed Event of a version is sent only once, but in special situations such as node failure and network partition, the Row Changed Event of the same version might be sent multiple times.
+* On the same table, the Row Changed Events of each version which is first sent are incremented in the order of timestamps (TS) in the Event stream.
+* Resolved Events are periodically broadcasted to each MQ Partition. The Resolved Event means that any Event with a TS earlier than Resolved Event TS has been sent to the downstream.
+* DDL Events are broadcasted to each MQ Partition.
+* Multiple Row Changed Events of a row are sent to the same MQ Partition.
+
+## Benchmark
+
+* Binary protocol encodes the same events with far less amount of bytes.
+* Protobuf is a popular binary format that can be used instead of craft format. Since craft format can apply sophisticated optimizations, it can be more compact and faster than protobuf.
+
+Serialized size:
+
+| case | craft size | json size | protobuf 1 size | protobuf 2 size | craft compressed | json compressed | protobuf 1 compressed | protobuf 2 compressed |
+| :---- | :--------- | :-------- | :-------------- | :-------------- | :--------------- | :-------------- | :-------------------- | :-------------------- |
+| case 0 | 180 | 390 (116%)+ | 197 (9%)+ | 199 (10%)+ | 136 | 190 (39%)+ | 157 (15%)+ | 150 (10)%+ |
+| case 1 | 722 | 1576 (118%)+ | 820 (13%)+ | 810 (12%)+ | 256 | 305 (19%)+ | 266 (3%)+ | 262 (2)%+ |
+
+Encoding speed:
+
+| craft | json | protobuf 1 | protobuf 2 |
+| :---- | :--- | :--------- | :--------- |
+| 2052 ns/op | 15326 ns/op (647%)+ | 2278 ns/op (11%)+ | 2746 ns/op (34%)+ |
+
+Decoding speed:
+
+| craft | json | protobuf 1 | protobuf 2 |
+| :---- | :--- | :--------- | :--------- |
+| 4227 ns/op | 40547 ns/op (859%)+ | 4089 ns/op (3%)- | 4670 ns/op (10%) |
+
+## Message format
+
+A Message contains one or more Events, arranged in the following format:
+
+Message:
+
+| uvarint | uvarint | header | body | size tables |
+| :------ | :------ | :------ | :------ | :------ |
+| version | number of events | events header | events body | size tables |
+
+Header:
+
+| delta uvarint chunk | uvarint chunk | uvarint chunk | varint chunk | string chunk | string chunk |
+| :------ | :------ | :------ | :------ | :------ | :------ |
+| commit ts | event type | row id | partition id (-1 for no partition) | schema | table |
+
+Body for [Row Changed Event](#row-changed-event):
+
+| Column group | Column group |
+| :------ | :------ |
+| column group 1 | column group 2 (optional) |
+
+Column group:
+| 1 byte | uvarint | string chunk | uvarint chunk | uvarint chunk | nullable bytes chunk |
+| :------ | :------ | :------ | :------ | :------ | :------ |
+| type: 1 New, 2 Old, 3 Delete | number of columns | name | type | flag | value |
```

Comment:
What does `type` mean here for a column?

Reply:
Each group is basically a row of data, type is used to tell user if this is the old value of given row or the new value otherwise.